src/core/expression/qgsexpressionutils.h

```diff
@@ -94,30 +95,32 @@ class CORE_EXPORT QgsExpressionUtils
         return Unknown;
 
       //handle some special cases
+      int userType = value.userType();
       if ( value.type() == QVariant::UserType )
       {
-        if ( value.userType() == qMetaTypeId< QgsGeometry>() )
+        if ( userType == qMetaTypeId< QgsGeometry>() || userType == qMetaTypeId<QgsReferencedGeometry>() )
         {
           //geom is false if empty
-          const QgsGeometry geom = value.value<QgsGeometry>();
+          const QgsGeometry geom = getGeometry( value, nullptr );
           return geom.isNull() ? False : True;
         }
-        else if ( value.userType() == qMetaTypeId<QgsFeature>() )
+        else if ( userType == qMetaTypeId<QgsFeature>() )
         {
           //feat is false if non-valid
           const QgsFeature feat = value.value<QgsFeature>();
           return feat.isValid() ? True : False;
         }
       }
 
-      if ( value.userType() == QMetaType::Type::Int )
+      if ( userType == QMetaType::Type::Int )
         return value.toInt() != 0 ? True : False;
 
       bool ok;
       const double x = value.toDouble( &ok );
       if ( !ok )
       {
-        parent->setEvalErrorString( QObject::tr( "Cannot convert '%1' to boolean" ).arg( value.toString() ) );
+        if ( parent != nullptr )
```

Comment:
QGIS coding convention is just to test pointers directly:

```suggestion
        if ( parent )
```